<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference • integration</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">integration</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.27</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="Integration-package.html">Integration-package</a></code> </p>
        </td>
        <td><p>Integrates data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="McKeonHomeogenity.html">McKeonHomeogenity()</a></code> </p>
        </td>
        <td><p>Calculates McKeon Homeogenity</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="access_reactome.html">access_reactome()</a></code> </p>
        </td>
        <td><p>Return data from reactome</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="allComb.html">allComb()</a></code> </p>
        </td>
        <td><p>Logical vectors of meta data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="analyze.html">analyze()</a></code> </p>
        </td>
        <td><p>Analyze a sgcca object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="angle.html">angle()</a></code> </p>
        </td>
        <td><p>Calculates the angle between to slopes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aves.html">aves()</a></code> </p>
        </td>
        <td><p>Method to simplify AVE</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="biological_relationships.html">biological_relationships()</a></code> </p>
        </td>
        <td><p>Performs the calculation of biological information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="boot_evaluate.html">boot_evaluate()</a></code> </p>
        </td>
        <td><p>Evaluates the bootstrapping of RGCCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="boot_sgcca.html">boot_sgcca()</a></code> </p>
        </td>
        <td><p>Bootstrap sgcca</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="boots_corr.html">boots_corr()</a></code> </p>
        </td>
        <td><p>Bootstrapping to asses how probable is to have such enrichment in those p-values matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="check_design.html">check_design()</a></code> </p>
        </td>
        <td><p>Validate designs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circle.html">circle</a></code> </p>
        </td>
        <td><p>A circle of radium 2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circleFun.html">circleFun()</a></code> </p>
        </td>
        <td><p>Create a circle</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clean_unvariable.html">clean_unvariable()</a></code> </p>
        </td>
        <td><p>Remove unvariable features</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="colors.html">colors</a></code> </p>
        </td>
        <td><p>Useful colors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="comb_prevalence.html">comb_prevalence()</a></code> </p>
        </td>
        <td><p>Look for prevalence in combinations of two</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="comp.html">comp()</a></code> </p>
        </td>
        <td><p>Analyze the data for the relationship in time</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare.html">compare()</a></code> </p>
        </td>
        <td><p>Compares two objects of class sgcca</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare.correlations.html">compare.correlations()</a></code> </p>
        </td>
        <td><p>Calculates the z-score of two correlations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="contingency_taxa.html">contingency_taxa()</a></code> </p>
        </td>
        <td><p>Compares the taxonomy of the otus</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convert.z.score.html">convert.z.score()</a></code> </p>
        </td>
        <td><p>Z-score to p-value</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cor_sign.html">cor_sign()</a></code> </p>
        </td>
        <td><p>Significative correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="correct.html">correct()</a></code> </p>
        </td>
        <td><p>Check that the network is fully connected</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cors_rgcca.html">cors_rgcca()</a></code> </p>
        </td>
        <td><p>Check the efficacy of RGCCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dist2d.html">dist2d()</a></code> </p>
        </td>
        <td><p>Calculate the distance between a line and a point</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ensembl2symbol.html">ensembl2symbol()</a></code> </p>
        </td>
        <td><p>Translate ensmbl to symbols</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="epitheliumE.html">epitheliumE()</a></code> </p>
        </td>
        <td><p>Function to read the epithelium signature</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fastercheck.html">fastercheck()</a></code> </p>
        </td>
        <td><p>Check if a vector is in the matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_RNAseq.html">filter_RNAseq()</a></code> </p>
        </td>
        <td><p>Filter expressions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filter_values.html">filter_values()</a></code> </p>
        </td>
        <td><p>Filter correlations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="full_prevalence.html">full_prevalence()</a></code> </p>
        </td>
        <td><p>Look for prevalence of a factor</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAVEs.html">getAVEs()</a></code> </p>
        </td>
        <td><p>Retrieve inner AVE</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="improve.sgcca.html">improve.sgcca()</a></code> </p>
        </td>
        <td><p>Improve the information on sgcca classes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="integration.html">integration()</a></code> </p>
        </td>
        <td><p>Integrate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="intercorrelation.html">intercorrelation()</a></code> </p>
        </td>
        <td><p>Intracorrelation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kFolding.html">kFolding()</a></code> </p>
        </td>
        <td><p>Indexes for K-folds</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="looIndex.html">looIndex()</a></code> </p>
        </td>
        <td><p>Indexes without one sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loo_functions.html">loo_functions()</a></code> </p>
        </td>
        <td><p>Function to export a function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="m_sem.html">m_sem()</a></code> </p>
        </td>
        <td><p>Summarize a model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeRects.html">makeRects()</a></code> </p>
        </td>
        <td><p>Make a rectangle at those cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meta_i_norm.html">meta_i_norm()</a></code> </p>
        </td>
        <td><p>Normalize 16S biopsies metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meta_r_norm.html">meta_r_norm()</a></code> </p>
        </td>
        <td><p>Normalize the metadata of the RNA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meta_s_norm.html">meta_s_norm()</a></code> </p>
        </td>
        <td><p>Normalize 16S stools metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="model_RGCCA.html">model_RGCCA()</a></code> </p>
        </td>
        <td><p>Prepare data for CCA.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="model_columns.html">model_columns()</a></code> </p>
        </td>
        <td><p>Adapt data for a CCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="norm_RNAseq.html">norm_RNAseq()</a></code> </p>
        </td>
        <td><p>Normalize RNAseq</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="norm_expr_colnames.html">norm_expr_colnames()</a></code> </p>
        </td>
        <td><p>Change names of expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="norm_otus.html">norm_otus()</a></code> </p>
        </td>
        <td><p>Normalize OTUS</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pathsPerMicro.html">pathsPerMicro()</a></code> </p>
        </td>
        <td><p>Enrichment by microorganisms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permanova_expr.html">permanova_expr()</a></code> </p>
        </td>
        <td><p>Permanova</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="permanova_otus.html">permanova_otus()</a></code> </p>
        </td>
        <td><p>Permanova</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotAVEs.html">plotAVEs()</a></code> </p>
        </td>
        <td><p>Distribution of inner AVE</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_cor.html">plot_cor()</a></code> </p>
        </td>
        <td><p>Plot a correlation with ggplot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_interesting.html">plot_interesting()</a></code> </p>
        </td>
        <td><p>Plot  PCA of interesting variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_samples.html">plot_samples()</a></code> </p>
        </td>
        <td><p>Plot samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_single_cor.html">plot_single_cor()</a></code> </p>
        </td>
        <td><p>Plot correlations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_variables.html">plot_variables()</a></code> </p>
        </td>
        <td><p>Plot bullseye</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prevalence.html">prevalence()</a></code> </p>
        </td>
        <td><p>Test prevalence</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prevalence_tab.html">prevalence_tab()</a></code> </p>
        </td>
        <td><p>Calculates the presence or absence of a microorganism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="probability_samples.html">probability_samples()</a></code> </p>
        </td>
        <td><p>Calculates the probability of obtaining these samples.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pvalue.html">pvalue()</a></code> </p>
        </td>
        <td><p>Calculates the p.value</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ratio.html">ratio()</a></code> </p>
        </td>
        <td><p>Ratio of prevalence</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="readSGCCA.html">readSGCCA()</a></code> </p>
        </td>
        <td><p>Read component</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="relevant.html">relevant()</a></code> </p>
        </td>
        <td><p>List the correlations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="selectVar.html">selectVar()</a></code> </p>
        </td>
        <td><p>Select variable from bootstrapping</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="select_genes_int.html">select_genes_int()</a></code> </p>
        </td>
        <td><p>Filter genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="select_var.html">select_var()</a></code> </p>
        </td>
        <td><p>Select important variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sem.html">sem()</a></code> </p>
        </td>
        <td><p>Standard error of the mean</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sign_cor.html">sign_cor()</a></code> </p>
        </td>
        <td><p>List the correlations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="size.html">size()</a></code> </p>
        </td>
        <td><p>Number of samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="store_micro.html">store_micro()</a></code> </p>
        </td>
        <td><p>Store result by microorganism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subSymm.html">subSymm()</a></code> </p>
        </td>
        <td><p>Substitute in a symmetric matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetData.html">subsetData()</a></code> </p>
        </td>
        <td><p>Subset a list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="symm.html">symm()</a></code> </p>
        </td>
        <td><p>Create symmetric matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="taxonomy.html">taxonomy()</a></code> </p>
        </td>
        <td><p>Clean and prepare the data from IMNGS</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tidyer.html">tidyer()</a></code> </p>
        </td>
        <td><p>Clean the output of a sgcca object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tol21rainbow.html">tol21rainbow</a></code> </p>
        </td>
        <td><p>Rainbow colors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trimVer.html">trimVer()</a></code> </p>
        </td>
        <td><p>Trim version number of genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="two.sided.html">two.sided()</a></code> </p>
        </td>
        <td><p>Two sided test</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="variables.html">variables()</a></code> </p>
        </td>
        <td><p>Extract in a tidy way the information about the variables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="variables_weight.html">variables_weight()</a></code> </p>
        </td>
        <td><p>Plot density of the weight of components</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="weight_design.html">weight_design()</a></code> </p>
        </td>
        <td><p>Check a design for different weights</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="weights.html">weights()</a></code> </p>
        </td>
        <td><p>Select the weights and adds information for human RNAseq</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="weights_bayes.html">weights_bayes()</a></code> </p>
        </td>
        <td><p>Independence between genes and OTUs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="weights_correlation.html">weights_correlation()</a></code> </p>
        </td>
        <td><p>Correlation weights</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="weights_otus.html">weights_otus()</a></code> </p>
        </td>
        <td><p>Select the weights and adds information for taxa RNAseq</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wilks_rgcca.html">wilks_rgcca()</a></code> </p>
        </td>
        <td><p>Check the efficacy of RGCCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write_cor.html">write_cor()</a></code> </p>
        </td>
        <td><p>Translate symbols and store</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Lluís Revilla Sancho.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.0.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


